CODINGSTYLE.md

```diff
@@ -0,0 +1,490 @@
+
+## Coding style for OpenTTD
+### Functions
+* Function names use [CamelCase](http://www.wikipedia.org/wiki/Camelcase) without underscores.
+* Opening curly bracket **{** for a function starts on the next line.
+* Use Foo() instead of Foo(void).
+```c++
+void ThisIsAFunction()
+{
+}
+```
+
+### Variables
+* Variable names are all lowercase, and use "_" as separator.
+* Global variables are preceded by an underscore. ("_") Use descriptive names because leading underscores are often used for system / compiler variables.
+* Own members of classes should always be referenced using "this->"
+* Pointers and references should have their reference symbol next to the name (compatibility with current code).
+* Variables that are declared below one another should have their type, name or reference operator, and assignment operator aligned by spaces.
+* There are set names for many variables. Those are (but not limited to): Vehicle *u, *v, *w; Station *st; Town *t; Window *w; Engine *e.
+* For multiple instances, use numbers "*t1, *t2" or postfixes "*st_from, *st_to".
+* Declare variables upon first usage.
+* Declare iterators in their loop.
+* There is a space between '*' and 'const' in "const pointers"
+```c++
+int     number         = 10;
+Vehicle *u_first_wagon = v->next;
+int     value          = v->value;
+
+uint32 _global_variable = 3750;
+
+static const char * const _global_array[] = {
+	"first string",
+	"second string",
+	"another string",
+	"last string followed by comma",
+};
+
+protected:
+	char protected_array[10];
+
+for (int i = 0;;);
+```
+
+* Give the variables expedient names, this increases the readability of the code
+```c++
+bool is_maglev_train = true;
+if (!is_maglev_train) DoSomething();
+```
+
+* Some people like to introduce copies of variables to increase readability, which can waste memory. However, in some cases, especially in code pieces which are often called, it makes sense to cache some calculated variables.
+```c++
+/* Unoptimized code:
+ * foo is not touched inside the loop!
+ */
+for (uint8 i = 0; i < 100000; i++) {
+	/* Do something */
+	if (value_to_check == (foo * 4) % 5 + 6) DoSomething();
+	/* Do something else */
+}
+
+/* Better:
+ * The used value of foo is calculated outside the loop.
+ */
+const uint32 bar = (foo * 4) % 5 + 6;
+for (uint8 i = 0; i < 100000; i++) {
+	/* Do something */
+	if (value_to_check == bar) DoSomething();
+	/* Do something else */
+}
+```
+
+### Enumerations / static consts
+* Enumerations store integers that belong logically together (railtypes, string IDs, etc.).
+* Enumeration names also use CamelCase.
+* Unscoped enumerators are all caps with "_" between the words.
+* Scoped enumerators are use CamelCase.
+* Enums are not used to store single numbers.
+* Enums have consecutive numbers OR
+* Enums have consecutive powers of two. Powers of two (bits) are written in hex or with the shift operator.
+* Enums may have special enumerators: "_BEGIN", "_END", and "INVALID_").  See example.
```

Comment:
When rendering the current variant in GitHub it makes 'END", and "INVALID' italic, and more importantly it removes the underscores. Maybe reordering them solves the issue?
```suggestion
* Enums may have special enumerators: "INVALID_", "_BEGIN", and "_END").  See example.
```